src/platforms/ctxlink/platform.c

```diff
@@ -0,0 +1,305 @@
+	//
+	// See the spreadsheet "SWD Port speed bits - OneNote"
+	//
+	GPIOA_OSPEEDR &= ~(TCK_PIN | TMS_PIN | TDI_PIN); // Clear the speed bits for TCK, TMS, & TDI
+	GPIOA_OSPEEDR |= (TCK_PIN | TMS_PIN | TDI_PIN);  // Set TCK, TMS,& TDI to "Fast speed"
+	gpio_mode_setup(JTAG_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, TMS_DIR_PIN | TMS_PIN | TCK_PIN | TDI_PIN);
+
+	gpio_mode_setup(TDO_PORT, GPIO_MODE_INPUT, GPIO_PUPD_NONE, TDO_PIN);
+	//
+	// Initialize the LED ports
+	//
+	gpio_mode_setup(LED_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, LED_IDLE_RUN | LED_ERROR | LED_MODE);
+	gpio_mode_setup(LED_PORT_UART, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, LED_UART);
+	//
+	// Setup RST_SENSE as input
+	//
+	//	Give it a pullup (NOT reset) just in case similar issue to
+	// native firmware.
+	//
+	gpio_mode_setup(NRST_SENSE_PORT, GPIO_MODE_INPUT, GPIO_PUPD_PULLUP, NRST_SENSE_PIN);
+	/* Enable nRST output. Original uses a NPN to pull down, so setting the
+	 * output HIGH asserts. Mini is directly connected so use open drain output
+	 * and set LOW to assert.
+	 */
+	platform_nrst_set_val(false);
+	//
+	// setup the iRSTR pin
+	//
+	gpio_mode_setup(NRST_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_PULLUP, NRST_PIN);
+	/*
+	 * Enable internal pull-up on PWR_BR so that we don't drive
+	 * TPWR locally or inadvertently supply power to the target.
+	 */
+	gpio_set(PWR_BR_PORT, PWR_BR_PIN);
+	gpio_mode_setup(PWR_BR_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, PWR_BR_PIN);
+	gpio_set_output_options(GPIOB, GPIO_OTYPE_OD, GPIO_OSPEED_50MHZ, PWR_BR_PIN);
+
+	platform_timing_init();
+	adc_init();
+	blackmagic_usb_init();
+	aux_serial_init();
+
+	/* https://github.com/libopencm3/libopencm3/pull/1256#issuecomment-779424001 */
+	OTG_FS_GCCFG |= OTG_GCCFG_NOVBUSSENS | OTG_GCCFG_PWRDWN;
+	OTG_FS_GCCFG &= ~(OTG_GCCFG_VBUSBSEN | OTG_GCCFG_VBUSASEN);
+
+	/* By default, do not drive the SWD bus too fast. */
+}
+
+void platform_nrst_set_val(bool assert)
+{
+	(void)assert;
+}
+
+bool platform_nrst_get_val(void)
+{
+	return false;
+}
+
+const char *platform_target_voltage(void)
+{
+	return NULL;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
+
+void platform_request_boot(void)
+{
+	const uint32_t application_address = 0x1fff0000;
+	register uint32_t jump_address = 0;
+	register uint32_t addr = 0;
+	static irq_function_t jump_to_application;
+
+	/* We start here */
+	cm_disable_interrupts();
+	uint32_t value = UINT32_MAX;
+	NVIC_ICER(0) = value;
+	NVIC_ICER(1) = value;
+	NVIC_ICER(2) = value;
+	NVIC_ICPR(0) = value;
+	NVIC_ICPR(1) = value;
+	NVIC_ICPR(2) = value;
+
+	STK_CSR = 0;
+	/* Reset the RCC clock configuration to the default reset state ------------*/
+	/* Reset value of 0x83 includes Set HSION bit */
+	RCC_CR |= (uint32_t)0x00000082;
+	/* Reset CFGR register */
+	RCC_CFGR = 0x00000000;
+
+	/* Disable all interrupts */
+	RCC_CIR = 0x00000000;
+
+	FLASH_ACR = 0;
+	__asm__ volatile("isb");
+	__asm__ volatile("dsb");
```

Comment:
Note that there are primitives for this in a different platform, stlinkv3, used in Cortex-M7 cache setup, also written differently:
https://github.com/blackmagic-debug/blackmagic/blob/bbc1ba51b7a152835e6fe7444416875d8249a320/src/platforms/stlinkv3/platform.c#L61-L73
I'd add DMB if MPU setup was required, per `<mpu_v7.h>`, but your usage seems to do the Jump to MaskROM without rebooting, which is also valid. If this works better for your platform than the double-reboot I implemented/debugged and tested for blackpill-f4, then nevermind. Does it exit from MaskROM bootloader into BMF firmware/application without problems?

Reply:
I just ran a firmware update to check the exit from bootloader to BMF and it requires a reset to run the BMF firmware.

Dealing with right now would be a distraction so I am going to note your comment and address it later, assuming, with your approach the reset is not required.
Thanks.

Reply:
Unfortunately this reset is also required in my experience. I'm likely missing some extra deinit/reset pokes on the USBOTG. Just so you know I might look into this or test suggestions. Yeah, get the USB-only platform in-tree, I won't distract further.

Reply:
Ping, see #1861 and report if that works for your, F401RE, I assume. It was unrelated to OTGFS.

Reply:
@ALTracer , I am confused by the comment in your patch. 

I am using the ST System bootloader, are you suggesting adding your path before jumping to the bootloader? If so, your comment says "Unmap ST MaskROM and map back to Internal Flash"

If I am jumping to the bootloader, why would I unmap it?
:)

Reply:
As best as we understand from the explanation and where the remap calls are made - it's when jumping back (ST bootloader -> BMD) so that when the firmware tries to engage interrupt handling, VTOR points at the right vector table.

Reply:
No-no, the recent development is that I **have to** unmap System bootloader when it exits back into Blackmagic Debug Firmware -- I have tested this to work well with either raw BMF or BMF+BMPBootloader (I like the latter because it flashes faster). Note that the patch adds code *after* any old jumping code. You still can enter the ST MaskROM bootloader via enabling Remap and branching (or resetting core), it's the way back that was broken -- the infamous hang and having to press reset button (if it even exists).
You can review that there if you have more questions.

Reply:
So, this needs to be done early in the ctxlink application code, is that correct?

Reply:
Correct. In the PR description I wrote why, but tl;dr is you have to do that after SYSCFG is clocked and before any interrupts are unblocked, because the first interrupt will vector you to 0 (or whatever SCB_VTOR) and there might be MaskROM sitting due to remap. `platform_init` is a good place for it, because BMF enables SysTick and USB and USART and whatnot.

Reply:
`	rcc_clock_setup_pll(&rcc_hse_8mhz_3v3[RCC_CLOCK_3V3_84MHZ]);
	/*
	 * Ensure the ST BOOTloader is unmapped
	 *
	 * Ref: https://github.com/blackmagic-debug/blackmagic/pull/1861
	 *
	 */
	if ((SYSCFG_MEMRM & 3U) == 1U)
		SYSCFG_MEMRM &= ~3U;

`
No, still have to press reset after bootloader.

Reply:
+`rcc_periph_clock_enable(RCC_SYSCFG);` or something. Per RM it might be unclocked, see default values of RCC_APB1EN register etc. You would be reading zeroes but it's actually 0b01.

Reply:
I added the rcc_periph_clock_enable and still not working.

Also, the value in SYSCFG_MEMRM should be 0b00, according to the TRM:

![syscfg_memrm](https://github.com/user-attachments/assets/c633a4dd-6a56-4a00-a7df-c25bf4ba792c)